windows/scripts/extract_kernels.py

```diff
@@ -0,0 +1,201 @@
+#!/usr/bin/env python
+
+# Copyright (C) 2017, Adrian Herrera
+# All rights reserved.
+#
+# Licensed under the MIT License.
+
+"""
+Extracts the Windows kernel executable (``ntoskrnl.exe``) from an ISO. This is
+useful when updating the ``winmonitor_gen.c`` file, as the kernel executable
+is used to generate the correct data structure offsets for that particular
+kernel.
+
+Usage
+-----
+
+::
+
+    python extract_kernels.py -d /path/to/isos -o /path/to/output
+
+Where ``isos`` is a directory containing all or some of the supported ISOs
+listed in s2e/guest-images/images.json and ``output`` is the directory to store
+the kernel executable files.
+
+Requirements
+------------
+
+Requires 7-Zip (http://www.7-zip.org/) and that the ``7z`` executable is
+available on the ``PATH``.
+
+This script will work with both Python 2.7 and 3.x.
+"""
+
+
+from __future__ import print_function
+
+
+import argparse
+import os
+import shutil
+import subprocess
+
+try:
+    from subprocess import DEVNULL
+except ImportError:
+    DEVNULL = open(os.devnull, 'wb')
+
+import sys
+import tempfile
+
+
+# Maps Windows ISOs (taken from the supported images listed in
+# s2e/guest-images/images.json) to the location of ntoskrnl.exe within the ISO.
+#
+# ntoskrnl.exe is stored within an intermediary container. The path to this
+# intermediary container is given in the first element in the value tuple. The
+# second tuple element is the path to ntoskrnl.exe within this intermediary
+# container.
+NTOSKRNL_MAP = {
+    'en_windows_xp_professional_with_service_pack_3_x86_cd_x14-80428.iso':
+        (os.path.join('I386', 'NTOSKRNL.EX_'),
+         'ntoskrnl.exe'),
+    'en_windows_7_enterprise_with_sp1_x64_dvd_u_677651.iso':
+        (os.path.join('sources', 'install.wim'),
+         os.path.join('Windows', 'System32', 'ntoskrnl.exe')),
+    'en_windows_8_1_enterprise_x64_dvd_2971902.iso':
```

Comment:
Does the list of ISOs we have cover the existing kernels? IIRC, it also covered Win 8.

I think it's better to not hard-code file names. Take an arg containing a dir with all ISOs, then go over each ISO and try to extract as much as possible from them.

Reply:
The ISOs match those listed in https://github.com/S2E/guest-images/blob/master/images.json

guest-images uses the same hard-coded file names, which is why I used them here.

Reply:
I meant those that we already support in https://github.com/S2E/guest-tools/blob/master/windows/driver/src/winmonitor_gen.c.
How many are we missing with the current images that we support?

Reply:
To clarify a bit more. Currently, winmonitor_gen.c supports more OS versions than what we have in the image build scripts. So, what's the intersection between the winmonitor_gen.c currently in the repo with the one you would get by running your scripts.

Reply:
Right, I understand now. I will look into it